internal/locCluster.ecl

```diff
@@ -0,0 +1,314 @@
+                            Union(ds,ds[i],neighs[n]);
+                        }
+                    } else {
+                        if(!neighs[n]->ifCore){
+                            if(getNeighbors(ds,neighs[n],eps).size()>=minpts){
+                                neighs[n]->ifCore = true;
+                                Union(ds,ds[i],neighs[n]);
+                            } else {
+                                if(neighs[n]->isVisited) continue;
+                                neighs[n]->isVisited = true;
+                                Union(ds,ds[i],neighs[n]);
+                            }
+                        } else {
+                            Union(ds,ds[i],neighs[n]);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // The data structure for the return layout.
+    struct retRecord{
+    uint32_t wi;
+    uint32_t id;
+    uint32_t parentId;
+    uint32_t nodeId;
+    bool if_local;
+    bool if_core;
+    };
+    //ResultStream Interface returns the resulting rows as a stream for global merge phase.
+    //Uses retRecord datastructure to store the results of local clustering
+    class ResultStream : public RtlCInterface, implements IRowStream {
+    public:
+        ResultStream(IEngineRowAllocator *_ra, IRowStream *_ds, uint64_t minpts, double eps,
+                                    unsigned long long lnode) : ra(_ra), ds(_ds), Lnode(lnode){
+            byte* p;
+            count = 0;
+            while((p=(byte*)ds->nextRow())){
+                node temp;
+                temp.wi = *((uint16_t*)p); p += sizeof(uint16_t);
+                temp.id = *((unsigned long long*)p); p += 2*sizeof(unsigned long long);
+                temp.nodeId = *((unsigned long long*)p); p += sizeof(unsigned long long);
+                p += sizeof(bool);
+                uint32_t lenFields = *((uint32_t*)p);
+                p += sizeof(uint32_t);
+                for(uint32_t i=0; i<lenFields/sizeof(float); ++i){
+                    float f = *((float*)p); p += sizeof(float);
+                    temp.fields.push_back(f);
+                }
+                temp.ifLocal = (lnode == temp.nodeId); p += sizeof(bool);
+                temp.ifCore = false;
+                dataset.push_back(temp);
+            }
+
+            for(uint64_t i=0; i<dataset.size(); ++i){
+                results.push_back(&dataset[i]);
+            }
+
+            dbscan(results, eps, minpts);
+        }
+        //Returning  row by row via Interface
+        RTLIMPLEMENT_IINTERFACE
+        virtual const void* nextRow() override {
+            RtlDynamicRowBuilder rowBuilder(ra);
+            if(count < results.size()){
+                uint32_t lenRec = 4*sizeof(uint32_t) + 2*sizeof(bool);
+                byte* p = (byte*)rowBuilder.ensureCapacity(lenRec, NULL);
+
+                while(!dataset[count].isModified && count < dataset.size())
```

Comment:
It is confusing that some tests refer to dataset and some to results.  It would be clearer if they all referred to dataset.  What is the reason for the results vector?

Reply:
It is because that the vector 'dataset' cannot directly feed into DBSCAN() function. It require a vector of pointers instead. In the test we use results to check against the count of dataset. I think we can somehow just use dataset in the test in the next stage.